crates/signs/src/line.rs

```diff
@@ -0,0 +1,268 @@
+enum LinesSet {
+    LineEnd,
+    LocationEvents,
+    VisibilityEvents,
+    SpawnLines,
+}
+
+#[derive(Resource)]
+struct LineMesh(Handle<Mesh>);
+
+fn setup(mut commands: Commands, mut meshes: ResMut<Assets<Mesh>>) {
+    commands.init_resource::<LineEntities>();
+    commands.init_resource::<LineTransforms>();
+    let line_mesh = meshes.add(shape::Plane::from_size(1.0).into());
+    commands.insert_resource(LineMesh(line_mesh));
+}
+
+fn cleanup(mut commands: Commands) {
+    commands.remove_resource::<LineEntities>();
+    commands.remove_resource::<LineTransforms>();
+    commands.remove_resource::<LineMesh>();
+}
+
+pub struct UpdateLineVisibilityEvent {
+    owner: Entity,
+    visible: bool,
+}
+
+impl UpdateLineVisibilityEvent {
+    pub fn new(owner: Entity, visible: bool) -> Self {
+        Self { owner, visible }
+    }
+}
+
+pub struct UpdateLineLocationEvent {
+    owner: Entity,
+    start: Vec3,
+    end: Vec3,
+}
+
+impl UpdateLineLocationEvent {
+    pub fn new(owner: Entity, start: Vec3, end: Vec3) -> Self {
+        Self { owner, start, end }
+    }
+}
+
+pub struct UpdateLineEndEvent {
+    owner: Entity,
+    end: Vec3,
+}
+
+impl UpdateLineEndEvent {
+    pub fn new(owner: Entity, end: Vec3) -> Self {
+        Self { owner, end }
+    }
+}
+
+struct DespawnLineEvent {
+    owner: Entity,
+}
+
+struct SpawnLineEvent(Transform, Entity);
+
+#[derive(Resource, Default)]
+struct LineEntities(AHashMap<Entity, Entity>);
+
+#[derive(Resource, Default)]
+struct LineTransforms(AHashMap<Entity, [Vec3; 2]>);
+
+fn update_line_end(
+    mut events: EventReader<UpdateLineEndEvent>,
+    lines: Res<LineTransforms>,
+    mut line_location: EventWriter<UpdateLineLocationEvent>,
+) {
+    for event in &mut events {
+        if let Some([start, _end]) = lines.0.get(&event.owner) {
+            line_location.send(UpdateLineLocationEvent::new(event.owner, *start, event.end));
+        }
+    }
+}
+
+fn update_line_location(
+    lines: Res<LineEntities>,
+    mut events: EventReader<UpdateLineLocationEvent>,
+    mut transforms: Query<&mut Transform>,
+    mut line_transforms: ResMut<LineTransforms>,
+) {
+    for event in &mut events {
+        let transform = Transform::from_matrix(compute_line_transform(event.start, event.end));
+        let positions = [event.start, event.end];
+        line_transforms.0.insert(event.owner, positions);
+        if let Some(line_entity) = lines.0.get(&event.owner) {
+            let mut current_transform = transforms.get_mut(*line_entity).unwrap();
+            *current_transform = transform
+        }
+    }
+}
+
+fn update_line_visibility(
+    mut events: EventReader<UpdateLineVisibilityEvent>,
+    mut lines: ResMut<LineEntities>,
+    line_transforms: Res<LineTransforms>,
+    mut spawn_line_events: EventWriter<SpawnLineEvent>,
+    mut despawn_line_events: EventWriter<DespawnLineEvent>,
+) {
+    for event in &mut events {
+        let line_entity = lines.0.entry(event.owner);
```

Comment:
Note that multiple events concerning single entity might be received during a game tick. This might lead to duplicate lines, etc. Some of the incorrect behaviour might not be possible due to how e.g. selection systems work currently, but to minimize assumptions about other systems / event sent, we should modify the implementation. I suggest the following:

All needed structs:

* `LineLocations` mapping owner `Entity` to a struct with `start`, `end`, `transform` fields.
* `LineEntities` mapping owner `Entity` to line `Entity`.

All needed systems:

* **setup**
* **cleanup**
* **update_line_end** - loads `start` from `LineLocations` and sends event to `update_line_location`
* **update_line_location** - Is executed after `update_line_end`. It updates `LineLocations` (with "upserts") by changing start/end and recomputing `transform`
* **update_line_visibility** - is executed after `update_line_location`. It spawns / despawns line entities based on `visibility`. It is conditioned on (non)existence of the owner in `LineEntities` (because multiple events of the same visibility might be sent in a row). It logs a warning if line is to be spawned for owner not present in `LineLocations`. It uses `transform` from `LineLocations` when spawning.
* **owner_despawn** - iterates over despawned factories. For each such factory it despawns line entity and removes the line from both `LineLocations` and `LineEntities`. It should run after `update_line_location` and `update_line_visibility`.